### PR TITLE
Kotlin: fix type variable erasure inside default function values

### PR DESCRIPTION
**Pull request by smowton**
Previously because extractClassInstance didn't use the declaration stack, we wouldn't notice that it was legal to refer to its type variable in the context of extracting a specialised method <-> method source-decl edge. This led to erasing the types of the source-decl, so that e.g. Map.put(...) would have signature (Object, Object) not (K, V) as it should.

### CODE REVIEW

File: java/ql/test/kotlin/library-tests/parameter-defaults/test.kt [hunk 1]
```diff
@@ -161,3 +161,15 @@ class VisibilityTests {
   private fun i(x: Int, y: Int = 0) = x + y
 
 }
+
+class TestGenericUsedWithinDefaultValue<T> {
+
+  // This tests parameter erasure works properly: we should notice that here the type variable T
+  // isn't used in the specialisation TestGenericUsedWithinDefaultValue<String>, but it can be
+  // cited in contexts like "the signature of the source declaration of 'TestGenericUsedWithinDefaultValue<String>.f(String)' is 'f(T)'",
+  // not 'f(Object)' as we might mistakenly conclude if we're inappropriately erasing 'T'.
```
Comment (tamasvajk) [inline java/ql/test/kotlin/library-tests/parameter-defaults/test.kt:170]: How would I see this in QL? Would this mean that `Callable.getSourceDeclaration().getSignature()` would return `ident(T)` on the target of `.ident("Hello world")`?

**Review by tamasvajk — COMMENTED** [review]
Looking at the PR, I'm unsure if the extractor changes are actually covered by the tests, are they?

File: java/ql/test/kotlin/library-tests/parameter-defaults/test.kt (hunk 1 above)
Comment (smowton) [reply java/ql/test/kotlin/library-tests/parameter-defaults/test.kt:170]: Yes. Without the extractor change we'd produce an inconsistency, trying to target a source-decl ident(Object) that doesn't exist -- you should be able to see it by reverting the .kt change and running the test.

**Review by tamasvajk — APPROVED** [review]
Looks plausible to me.

**Pull request merged**